rs/sns/governance/src/governance.rs

```diff
@@ -5079,55 +5079,70 @@ impl Governance {
-        let Some(deployed_version) = self.proto.deployed_version.as_ref() else {
-            log!(
-                ERROR,
-                "Cannot refresh cached_upgrade_steps: deployed_version not set."
-            );
-            return;
+    pub fn try_temporarily_lock_refresh_cached_upgrade_steps(&mut self) -> Result<Version, String> {
+        let deployed_version = self
+            .proto
+            .deployed_version
+            .clone()
+            .ok_or("Cannot lock refresh_cached_upgrade_steps: deployed_version not set.")?;
+
+        let mut cached_upgrade_steps = match self.proto.cached_upgrade_steps_or_err() {
+            Ok(cached_upgrade_steps) => cached_upgrade_steps,
+            Err(message) => {
+                self.push_to_upgrade_journal(upgrade_journal_entry::UpgradeStepsReset::new(
+                    message,
+                    vec![deployed_version.clone()],
+                ));
+                CachedUpgradeSteps::without_pending_upgrades(
+                    deployed_version.clone(),
+                    self.env.now(),
+                )
+            }
         };
+
+        // Lock the upgrade mechanism.
+        cached_upgrade_steps.requested_timestamp_seconds = self.env.now();
+        let cached_upgrade_steps = CachedUpgradeStepsPb::from(cached_upgrade_steps);
+        self.proto.cached_upgrade_steps = Some(cached_upgrade_steps);
+
+        Ok(deployed_version)
+    }
+
+    /// Refreshes the cached_upgrade_steps field
+    pub async fn refresh_cached_upgrade_steps(&mut self, deployed_version: Version) {
         let sns_governance_canister_id = self.env.canister_id().get();
 
         let upgrade_steps = crate::sns_upgrade::get_upgrade_steps(
             &*self.env,
-            deployed_version.clone(),
+            deployed_version,
             sns_governance_canister_id,
         )
         .await;
 
         let upgrade_steps = match upgrade_steps {
             Ok(upgrade_steps) => upgrade_steps,
             Err(err) => {
-                log!(
-                    ERROR,
-                    "Cannot refresh cached_upgrade_steps: call to SNS-W failed: {}",
-                    err
-                );
+                log!(ERROR, "Cannot refresh cached_upgrade_steps: {}", err);
                 return;
             }
         };
-        let upgrade_steps = Versions {
-            versions: upgrade_steps,
-        };
 
-        // Ensure `cached_upgrade_steps` is initialized
-        let cached_upgrade_steps = self
+        let new_cache = CachedUpgradeStepsPb::from(upgrade_steps);
+
+        let identical_upgrade_steps = self
             .proto
             .cached_upgrade_steps
-            .get_or_insert_with(Default::default);
-
-        // Update `response_timestamp_seconds`
-        cached_upgrade_steps.response_timestamp_seconds = Some(self.env.now());
+            .as_ref()
+            .map(|cache| cache.upgrade_steps == new_cache.upgrade_steps)
+            .unwrap_or_default();
```

Comment:
Previously this check was just `cached_upgrade_steps.upgrade_steps != Some(upgrade_steps.clone())`, but now it seems much more complicated? What is the reason for the increase in complexity?

Reply:
The new code avoids cloning. 

I'll see if I can refactor it for better readability.